### ISSUE TITLE
[ban-types] allow empty object {} type in intersection/union types

### ISSUE THREAD

**Issue by osdiab** [typescript-eslint/typescript-eslint#2884]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/ban-types": ["error"]
  }
}
```

```TS
// this is a legit usage of {} in my opinion, but ban-types excludes this possibility
const myIntersection: {a: string} & ({b: number, c: boolean} | {}) = {a: "hi"}
```

**Expected Result**

I think this is usage of {} has a real purpose - either b and c are both present or they're both not - but I can't express this without disabling the ban-types rule.

**Actual Result**

error error bleep bloop

**Additional Info**

idk how hard it is to implement this but I think that if `{}` is being used in an intersection or union type I'd love to be able to allow it with `ban-types` on.

Related to https://github.com/typescript-eslint/typescript-eslint/issues/2063#issuecomment-675156492 , where a similar case was mentioned re intersection types

**Comment by bradzacher** [comment]
Ban types is a simple rule. It does not contextually check anything. It simply bans a type.

It is not intended to be more advanced and complex. It is not going to receive special case logic for specific configurations.

It just bans a type from being used.

If you have what you believe to be a valid usage of a banned type, you can use a disable comment.
